packages/Uploader/src/useProcessFiles.js

```diff
@@ -91,90 +92,108 @@ export default function useProcessFiles({
-            file.status = types.SUCCESS;
+            file.key = uuidv4(); // change its key so it will be restartable
+            file.progress = 0;
+            file.request._data = undefined; // so superagent allows the upload to restart
+            file.request._aborted = false; // so superagent allows the upload to restart
             return fileItem;
           })
         );
-
-        areAllFilesProccessed();
       }
+    }
+  }
 
-      function onError({ file, error }) {
-        setFiles(
-          setFile(file, fileItem => {
-            const file = fileItem;
-            file.status = types.ERROR;
-            file.error = error;
-            file.processed = true;
-            file.isValid = false;
-            file.isServerValid = false;
-            return fileItem;
-          })
-        );
+  const upload = React.useCallback(() => {
+    const isSameList = (files, uploadingFileList) => {
+      const sameList =
+        files.length === uploadingFileList.length &&
+        files.every((file, index) => file.key === uploadingFileList[index].key);
+
+      return sameList;
+    };
 
-        areAllFilesProccessed();
+    function areAllFilesProccessed() {
+      if (files.every(file => file.processed)) {
+        setIsDisabled(() => false);
+        setisCompleted(() => true);
+        onCompleted(files);
       }
+    }
 
-      function onProgress({ file, percent }) {
-        setFiles(
-          setFile(file, fileItem => {
-            const file = fileItem;
-            file.progress = percent;
-            file.status = types.PROCESSING;
+    function onSuccess({ file, response }) {
+      setFiles(
+        setFile(file, fileItem => {
+          const file = fileItem;
+          file.isValid = true;
+          file.processed = true;
+          file.response = response;
+          file.status = statuses.SUCCESS;
+          return fileItem;
+        })
+      );
+
+      areAllFilesProccessed();
+    }
 
-            if (percent === 100) {
-              file.status = types.WAITINGFORSERVER;
-            }
-            return fileItem;
-          })
-        );
-      }
+    function onError({ file, error }) {
+      setFiles(
+        setFile(file, fileItem => {
+          const file = fileItem;
+          file.status = statuses.ERROR;
+          file.error = error;
+          file.processed = true;
+          file.isValid = false;
+          file.isServerValid = false;
+          return fileItem;
+        })
+      );
+
+      areAllFilesProccessed();
+    }
 
-      if (files.length && !isSameList(files, uploadingFileList) && !isDisabled) {
-        setUploadingFileList(() => files);
-        setIsDisabled(() => true);
-        setisCompleted(() => null);
-        onChange(files);
-        files.forEach(file => {
-          if (file.isValid && file.status !== types.SUCCESS) {
-            uploadToServer({ file, endpoint, onProgress, onSuccess, onError, headers });
+    function onProgress({ file, percent }) {
```

Comment:
shouldn't it be `onProgressChange`?

Reply:
If you wanna update function names, maybe `upload` → `onUpload` and `restartFileUpload` → `onRestartUpload`?

Reply:
Hmm not sure how i feel about these renaming suggestions.
`onProgress` goes with `onSuccess` and `onError`.

And when I want the upload to start, it doesn't feel right to fire `onUpload();`. Same for restarting.

Thoughts?